### PR TITLE
Offer default transaction timeout in Narayana

### PR DESCRIPTION
**Pull request by loicmathieu**
This PR allow to configure the default transaction timeout of Narayana.

It follows the discussion on the mailing list that is originated from this SO question : https://stackoverflow.com/questions/56746385/how-to-increase-transaction-timeout-in-quarkus/56750944

### CODE REVIEW

File: extensions/narayana-jta/runtime/src/main/java/io/quarkus/narayana/jta/runtime/NarayanaJtaConfiguration.java [hunk 1]
```diff
@@ -22,4 +22,10 @@
      */
     @ConfigItem()
     public Optional<String> xaNodeName;
+
+    /**
+     * The default transaction timeout is seconds
+     */
+    @ConfigItem
+    public Optional<Integer> defaultTransactionTimeout;
```
Comment (gsmet) [inline extensions/narayana-jta/runtime/src/main/java/io/quarkus/narayana/jta/runtime/NarayanaJtaConfiguration.java]: It should be a `Duration`.

And I think I would define the default value here so that it gets documented when we generate the doc.

File: docs/src/main/asciidoc/transaction-guide.adoc [hunk 2]
```diff
@@ -166,6 +166,11 @@ public class SantaClausService {
 You cannot use `UserTransaction` in a method having a transaction started by a `@Transactional` call.
 ====
 
+== Configuring the transaction timeout
+You can configure the default transaction timeout, the timeout that will applies to all transactions managed by the transaction manager, via the property `quarkus.narayana.default-transaction-timeout` (in seconds). The default value is unspecified in Quarkus, this means that the default of the transaction manager implementation is used.
+
+Currently, the transaction manager is Narayana that have a default of 60s.
```
Comment (gsmet) [inline docs/src/main/asciidoc/transaction-guide.adoc]: See my comment below about defining the default in Quarkus We could simplify the wording then.

File: docs/src/main/asciidoc/transaction-guide.adoc [hunk 3]
```diff
@@ -166,6 +166,11 @@ public class SantaClausService {
 You cannot use `UserTransaction` in a method having a transaction started by a `@Transactional` call.
 ====
 
+== Configuring the transaction timeout
+You can configure the default transaction timeout, the timeout that will applies to all transactions managed by the transaction manager, via the property `quarkus.narayana.default-transaction-timeout` (in seconds). The default value is unspecified in Quarkus, this means that the default of the transaction manager implementation is used.
```
Comment (gsmet) [inline docs/src/main/asciidoc/transaction-guide.adoc]:
```suggestion
You can configure the default transaction timeout, the timeout that will apply to all transactions managed by the transaction manager, via the property `quarkus.narayana.default-transaction-timeout` (in seconds). The default value is unspecified in Quarkus, this means that the default of the transaction manager implementation is used.
```

Comment (gsmet) [reply docs/src/main/asciidoc/transaction-guide.adoc]: It will get reorganized but, just for future work: add a new line at the end of each sentence. Asciidoctor will make it a single paragraph anyway and it makes diffs more readable.

File: extensions/narayana-jta/runtime/src/main/java/io/quarkus/narayana/jta/runtime/NarayanaJtaConfiguration.java [hunk 4]
```diff
@@ -9,7 +9,7 @@
 /**
  *
  */
-@ConfigRoot(phase = ConfigPhase.RUN_TIME)
+@ConfigRoot(name = "narayana", phase = ConfigPhase.RUN_TIME)
```
Comment (gsmet) [inline extensions/narayana-jta/runtime/src/main/java/io/quarkus/narayana/jta/runtime/NarayanaJtaConfiguration.java]: Maybe we should rename the configuration object instead.

File: extensions/narayana-jta/runtime/src/main/java/io/quarkus/narayana/jta/runtime/NarayanaJtaConfiguration.java [hunk 5]
```diff
@@ -22,4 +22,10 @@
      */
     @ConfigItem()
     public Optional<String> xaNodeName;
+
+    /**
+     * The default transaction timeout is seconds
```
Comment (gsmet) [inline extensions/narayana-jta/runtime/src/main/java/io/quarkus/narayana/jta/runtime/NarayanaJtaConfiguration.java]:
```suggestion
     * The default transaction timeout.
```

**Review by gsmet — COMMENTED** [review]
Added a few comments inline.

File: extensions/narayana-jta/runtime/src/main/java/io/quarkus/narayana/jta/runtime/NarayanaJtaConfiguration.java (hunk 4 above)
Comment (loicmathieu) [reply extensions/narayana-jta/runtime/src/main/java/io/quarkus/narayana/jta/runtime/NarayanaJtaConfiguration.java]: `transaction-manager` ? or just `transaction`

Comment (gsmet) [reply extensions/narayana-jta/runtime/src/main/java/io/quarkus/narayana/jta/runtime/NarayanaJtaConfiguration.java]: Not sure. We decided to expose Agroal configuration as `datasource` so maybe it makes sense to push this one as `transaction-manager` as Narayana is probably not very well known outside of the WF/EAP world.

@emmanuelbernard @FroMage thoughts?

Comment (emmanuelbernard) [reply extensions/narayana-jta/runtime/src/main/java/io/quarkus/narayana/jta/runtime/NarayanaJtaConfiguration.java]: makes sense

**Comment by emmanuelbernard** [comment]
Is there an issue for that feature?

File: extensions/narayana-jta/runtime/src/main/java/io/quarkus/narayana/jta/runtime/NarayanaJtaConfiguration.java (hunk 1 above)
Comment (loicmathieu) [reply extensions/narayana-jta/runtime/src/main/java/io/quarkus/narayana/jta/runtime/NarayanaJtaConfiguration.java]: OK, should I re-use the default from Narayana (60s) or configure a more higher default as we often find on some implementation (300s).

**Comment by loicmathieu** [comment]
@emmanuelbernard no, I can create one if you want.

File: extensions/narayana-jta/runtime/src/main/java/io/quarkus/narayana/jta/runtime/NarayanaJtaConfiguration.java (hunk 5 above)
Comment (loicmathieu) [reply extensions/narayana-jta/runtime/src/main/java/io/quarkus/narayana/jta/runtime/NarayanaJtaConfiguration.java]: done

File: extensions/narayana-jta/runtime/src/main/java/io/quarkus/narayana/jta/runtime/NarayanaJtaConfiguration.java (hunk 4 above)
Comment (loicmathieu) [reply extensions/narayana-jta/runtime/src/main/java/io/quarkus/narayana/jta/runtime/NarayanaJtaConfiguration.java]: done

File: docs/src/main/asciidoc/transaction-guide.adoc (hunk 3 above)
Comment (loicmathieu) [reply docs/src/main/asciidoc/transaction-guide.adoc]: rephrased

File: docs/src/main/asciidoc/transaction-guide.adoc (hunk 2 above)
Comment (loicmathieu) [reply docs/src/main/asciidoc/transaction-guide.adoc]: done

**Comment by loicmathieu** [comment]
@gsmet I move to timeout in duration and rephrase the documentation part.

File: docs/src/main/asciidoc/transaction-guide.adoc [hunk 6]
```diff
@@ -166,6 +166,21 @@ public class SantaClausService {
 You cannot use `UserTransaction` in a method having a transaction started by a `@Transactional` call.
 ====
 
+== Configuring the transaction timeout
+You can configure the default transaction timeout, the timeout that will applies to all transactions managed by the transaction manager, via the property `quarkus.transaction-manager.default-transaction-timeout`, specified as a duration.
```
Comment (gsmet) [inline docs/src/main/asciidoc/transaction-guide.adoc]:
```suggestion
You can configure the default transaction timeout, the timeout that applies to all transactions managed by the transaction manager, via the property `quarkus.transaction-manager.default-transaction-timeout`, specified as a duration.
```

File: docs/src/main/asciidoc/transaction-guide.adoc [hunk 7]
```diff
@@ -166,6 +166,21 @@ public class SantaClausService {
...
+
+[NOTE]
+====
+The format for durations uses the standard `java.time.Duration` format.
+You can learn more about it in the link:https://docs.oracle.com/javase/8/docs/api/java/time/Duration.html#parse-java.lang.CharSequence-[Duration#parse() javadoc].
+
+You can also provide duration values starting with a number.
+In this case, if the value consists only of a number, the converter treats the value as seconds.
+Otherwise, `PT` is implicitly appended to the value to obtain a standard `java.time.Duration` format.
+====
```
Comment (gsmet) [inline docs/src/main/asciidoc/transaction-guide.adoc]: maybe we should have an include for that one? I think we already have it in several places.

`duration-format-note.adoc`

File: extensions/narayana-jta/runtime/src/main/java/io/quarkus/narayana/jta/runtime/NarayanaJtaConfiguration.java [hunk 8]
```diff
@@ -9,7 +10,7 @@
 /**
  *
  */
-@ConfigRoot(phase = ConfigPhase.RUN_TIME)
+@ConfigRoot(name = "transaction-manager", phase = ConfigPhase.RUN_TIME)
 public final class NarayanaJtaConfiguration {
```
Comment (gsmet) [inline extensions/narayana-jta/runtime/src/main/java/io/quarkus/narayana/jta/runtime/NarayanaJtaConfiguration.java]: Let's rename the class to `TransactionManagerConfiguration` instead of specifying a name?

**Review by gsmet — CHANGES_REQUESTED** [review]
Added some more comments.

File: docs/src/main/asciidoc/transaction-guide.adoc (hunk 7 above)
Comment (gsmet) [reply docs/src/main/asciidoc/transaction-guide.adoc]: Note: this part can be done in a follow-up commit, let's focus on what is important here.

File: extensions/narayana-jta/runtime/src/main/java/io/quarkus/narayana/jta/runtime/NarayanaJtaConfiguration.java (hunk 1 above)
Comment (emmanuelbernard) [reply extensions/narayana-jta/runtime/src/main/java/io/quarkus/narayana/jta/runtime/NarayanaJtaConfiguration.java]: To me 60 seems fine but ask  on quarkus-dev so Michael can step up.

File: extensions/narayana-jta/runtime/src/main/java/io/quarkus/narayana/jta/runtime/NarayanaJtaConfiguration.java (hunk 8 above)
Comment (loicmathieu) [reply extensions/narayana-jta/runtime/src/main/java/io/quarkus/narayana/jta/runtime/NarayanaJtaConfiguration.java]: OK, done

File: docs/src/main/asciidoc/transaction-guide.adoc (hunk 7 above)
Comment (loicmathieu) [reply docs/src/main/asciidoc/transaction-guide.adoc]: There is no duration-format-note.adoc. 
I copy/paste the not from the Datasource documentation.

Comment (gsmet) [reply docs/src/main/asciidoc/transaction-guide.adoc]: Yes, sorry, I wasn't clear: it was about creating it to avoid duplication as I suspect we will need to have this note in several files (it's probably already the case).

But let's do that in another PR, once this one is in.

**Review by gsmet — APPROVED** [review]

**Comment by gsmet** [comment]
Merged, thanks!

File: docs/src/main/asciidoc/transaction-guide.adoc (hunk 7 above)
Comment (machi1990) [reply docs/src/main/asciidoc/transaction-guide.adoc]: I'll make a PR about the `duration-format-note.adoc` :-)
